ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,414 @@
+    FindVnodeRootFromDiskAndUpdateCache(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        context,
+        vnode,
+        vnodeHashIndex,
+        vnodeVid,
+        UpdateCacheBehavior_TrustCurrentEntry,
+        rootHandle);
+    
+    return rootHandle;
+}
+
+VirtualizationRootHandle VnodeCache_RefreshRootForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheHitCounter,
+    PrjFSPerfCounter cacheMissCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    vfs_context_t _Nonnull context)
+{
+    VirtualizationRootHandle rootHandle = RootHandle_None;
+    uintptr_t vnodeHashIndex = ComputeVnodeHashIndex(vnode);
+    uint32_t vnodeVid = vnode_vid(vnode);
+    
+    perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+    FindVnodeRootFromDiskAndUpdateCache(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        context,
+        vnode,
+        vnodeHashIndex,
+        vnodeVid,
+        UpdateCacheBehavior_ForceRefresh,
+        rootHandle);
+    
+    return rootHandle;
+}
+
+VirtualizationRootHandle VnodeCache_InvalidateVnodeRootAndGetLatestRoot(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheHitCounter,
+    PrjFSPerfCounter cacheMissCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    vfs_context_t _Nonnull context)
+{
+    VirtualizationRootHandle rootHandle = RootHandle_None;
+    uintptr_t vnodeHashIndex = ComputeVnodeHashIndex(vnode);
+    uint32_t vnodeVid = vnode_vid(vnode);
+    
+    perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+    FindVnodeRootFromDiskAndUpdateCache(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        context,
+        vnode,
+        vnodeHashIndex,
+        vnodeVid,
+        UpdateCacheBehavior_InvalidateEntry,
+        rootHandle);
+    
+    return rootHandle;
+}
+
+void VnodeCache_InvalidateCache(PerfTracer* _Nonnull perfTracer)
+{
+    perfTracer->IncrementCount(PrjFSPerfCounter_CacheInvalidateCount, true /*ignoreSampling*/);
+
+    RWLock_AcquireExclusive(s_entriesLock);
+    {
+        InvalidateCache_ExclusiveLocked();
+    }
+    RWLock_ReleaseExclusive(s_entriesLock);
+}
+
+KEXT_STATIC_INLINE void InvalidateCache_ExclusiveLocked()
+{
+    memset(s_entries, 0, s_entriesCapacity * sizeof(VnodeCacheEntry));
+}
+
+KEXT_STATIC_INLINE uint32_t ComputePow2CacheCapacity(int expectedVnodeCount)
+{
+    uint32_t idealCacheSize = expectedVnodeCount * 2;
+    uint32_t capacity = MinPow2VnodeCacheCapacity;
+    while (capacity < idealCacheSize && capacity < MaxPow2VnodeCacheCapacity)
```

Comment:
To confirm my understanding we set the capacity to the min.
We then shift "<< 1" until we are > than MaxPow2VnodeCacheCapacity?

Reply:
Yes that's correct, I'll clean this up a bit more an add some comments.